cynet/client.go

```diff
@@ -0,0 +1,582 @@
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("authentication failed with status %d: %s", resp.StatusCode, string(body))
+	}
+
+	var authResp AuthResponse
+	if err := json.Unmarshal(body, &authResp); err != nil {
+		return fmt.Errorf("failed to unmarshal auth response: %w", err)
+	}
+
+	if authResp.AccessToken == "" {
+		return fmt.Errorf("received empty access token from auth endpoint")
+	}
+
+	a.accessToken = authResp.AccessToken
+
+	return nil
+}
+
+func (a *CynetAdapter) fetchEvents() {
+	defer func() {
+		a.fetchOnce.Do(func() { close(a.chFetchLoop) })
+	}()
+
+	ticker := time.NewTicker(queryInterval * time.Second)
+	defer ticker.Stop()
+
+	// Initial fetch
+	a.runFetchCycle()
+
+	for {
+		select {
+		case <-a.ctx.Done():
+			return
+		case <-ticker.C:
+			a.runFetchCycle()
+		}
+	}
+}
+
+func (a *CynetAdapter) runFetchCycle() {
+	// a.conf.ClientOptions.DebugLog("starting fetch cycle")
+
+	// Fetch alerts
+	alerts, err := a.fetchAlerts()
+	if err != nil {
+		a.conf.ClientOptions.OnError(fmt.Errorf("failed to fetch alerts: %w", err))
+		a.consecutiveErrors++
+		if a.consecutiveErrors >= maxConsecutiveErrors {
+			a.conf.ClientOptions.OnError(fmt.Errorf("shutting down after %d consecutive fetch failures", maxConsecutiveErrors))
+			go a.Close()
+			return
+		}
+		return
+	}
+
+	// Reset error count on successful fetch
+	a.consecutiveErrors = 0
+
+	// Submit events
+	if len(alerts) > 0 {
+		//a.conf.ClientOptions.DebugLog(fmt.Sprintf("fetched %d new alerts", len(alerts)))
+		a.submitEvents(alerts)
+	} else {
+		//a.conf.ClientOptions.DebugLog("no new alerts found")
+	}
+}
+
+func (a *CynetAdapter) fetchAlerts() ([]utils.Dict, error) {
+	var allAlerts []utils.Dict
+	var latestTimestamp time.Time
+	var offset int
+
+	for {
+		select {
+		case <-a.ctx.Done():
+			return nil, a.ctx.Err()
+		default:
+		}
+
+		// Clean up old dedupe entries
+		a.cleanupDedupe()
+
+		// Build URL with pagination
+		u, err := url.Parse(fmt.Sprintf("%s%s", a.conf.URL, alertsEndpoint))
+
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse alerts URL: %w", err)
+		}
+
+		q := u.Query()
+
+		//a.conf.ClientOptions.DebugLog(fmt.Sprintf("Since: %s", a.since.Format("2006-01-02 15:04:05")))
+		q.Set("LastSeen", a.since.Format("2006-01-02 15:04:05"))
+		q.Set("Limit", strconv.Itoa(pageSize))
+		q.Set("Offset", strconv.Itoa(offset))
+		u.RawQuery = q.Encode()
+
+		// Make request
+		resp, err := a.makeRequest(u.String())
+		if err != nil {
+			return nil, err
+		}
+
+		// Process alerts
+		newAlerts, latest := a.processAlerts(resp)
+		allAlerts = append(allAlerts, newAlerts...)
+
+		if latest.After(latestTimestamp) {
+			latestTimestamp = latest
+		}
+
+		// Check if there are more pages
+		if len(resp.Entities) < pageSize || len(resp.Entities) == 0 {
+			break
+		}
+
+		offset += len(resp.Entities)
+
+		// Rate limiting between pages
+		select {
+		case <-a.ctx.Done():
+			return nil, a.ctx.Err()
+		case <-time.After(100 * time.Millisecond):
```

Comment:
`defer` inside a loop causes resource leak during retries. The response body should be closed explicitly before `continue`:
```go
body, err := io.ReadAll(resp.Body)
resp.Body.Close() // Close immediately, not deferred
```